File: Algorithms/Python/623 Add One Row to Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def addOneRow(self, root: TreeNode, v: int, d: int) -> TreeNode:
        if d == 1:
            node = TreeNode(v)
            node.left = root
            return node
        
        self.add_value_at(root, 1, v, d)
        return root
    
    def add_value_at(self, node, D, v, d):
        # current depth is D
        if D + 1 == d:
            left_node = node.left
            right_node = node.right
            new_left_node = TreeNode(v)
            new_right_node = TreeNode(v)
            node.left = new_left_node
            node.right = new_right_node
            new_left_node.left = left_node
            new_right_node.right = right_node
            return
        
        if node.left is not None:
            self.add_value_at(node.left, D + 1, v, d)

        if node.right is not None:
            self.add_value_at(node.right, D + 1, v, d)

File: Algorithms/C/623 Add One Row to Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

void add_one_row_at(struct TreeNode* node, int D, int v, int d)
{
    if (D + 1 == d){
        struct TreeNode* pLeft = node->left;
        struct TreeNode* pRight = node->right;
        struct TreeNode* newLeft = (struct TreeNode*)malloc(sizeof(struct TreeNode));
        struct TreeNode* newRight = (struct TreeNode*)malloc(sizeof(struct TreeNode));
        newLeft->val = v;
        newLeft->left = pLeft;
        newLeft->right = NULL;
        newRight->val = v;
        newRight->left = NULL;
        newRight->right = pRight;
        node->left = newLeft;
        node->right = newRight;
        return;
    }

    if (node->left){
        add_one_row_at(node->left, D + 1, v, d);
    }

    if (node->right){
        add_one_row_at(node->right, D + 1, v, d);
    }
}

struct TreeNode* addOneRow(struct TreeNode* root, int v, int d){
    if (d == 1){
        struct TreeNode* node = (struct TreeNode*)malloc(sizeof(struct TreeNode));
        node->val = v;
        node->left = root;
        node->right = NULL;
        return node;
    }

    add_one_row_at(root, 1, v, d);
    return root;
}
